### ISSUE TITLE
Bug: HDF5 1.10.8/1.13.0 installs CMake packages in wrong location

### ISSUE THREAD

**Issue by ax3l** [HDFGroup/hdf5#1320]
At least for Unix, HDF5 1.10.8/1.13.0 installs the CMake packages in the wrong location.

Please see @jbigot report in https://github.com/spack/spack/issues/28074 for details.

Note that Spack installs HDF5 via its CMake installation scripts: https://github.com/spack/spack/tree/develop/var/spack/repos/builtin/packages/hdf5

**Comment by ax3l** [comment]
Quoting @jbigot's Spack issue:

```console
spack install hdf5@1.10.8
ls ${SPACK_ROOT}/opt/spack/*/*/hdf5-*/share/cmake
```
output should be:
```
hdf5
```
as it is for 1.12.1 for example, but instead for 1.10.8, it is:
```
hdf5-config.cmake  hdf5-config-version.cmake  hdf5-targets.cmake  hdf5-targets-relwithdebinfo.cmake
```

As specified in cmake doc https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure
`<prefix>/share/cmake/` is not part of the Search Procedure, `<prefix>/share/cmake/<name>` is.

**Comment by ax3l** [comment]
Tracked the PRs that caused this:
- `develop` & `1.13.0` release: #1041
- `1.10.8` release: #1043

cc @byrnHDF :)

**Comment by ax3l** [comment]
This probably also affects the 1.13.0 release

**Comment by byrnHDF** [comment]
<prefix>/(lib/<arch>|lib*|share)/cmake/<name>*/                 (U)
is part of the search path for *nix.

**Comment by jbigot** [comment]
Dear @byrnHDF ,

in [CMake doc](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure), the 3 lines I see for Unix are:
```
<prefix>/(lib/<arch>|lib*|share)/cmake/<name>*/                 (U)
<prefix>/(lib/<arch>|lib*|share)/<name>*/                       (U)
<prefix>/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/         (U)
```
plus 3 common lines for Win/Unix
```
<prefix>/<name>*/(lib/<arch>|lib*|share)/cmake/<name>*/         (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/               (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/ (W/U)
```

All these 6 lines contain `<name>` in one form or another

**Comment by byrnHDF** [comment]
>> All these 6 lines contain `<name>` in one form or another

They contain `<name>*` - so for hdf5 the files are:
hdf5-config.cmake  hdf5-config-version.cmake  hdf5-targets.cmake  hdf5-targets-relwithdebinfo.cmake

**Comment by byrnHDF** [comment]
HDF5 CMake mininum is 3.12 and introduced in 3.12 is the <NAME>_ROOT define (CMake or ENV VAR). Setting that variable to the install location of hdf5 will allow the find package process to find the config files whereever they are located.

https://cmake.org/cmake/help/latest/command/find_package.html
----------------------------------------------------------------------------------------
New in version 3.12: Search paths specified in the <PackageName>_ROOT CMake variable and the <PackageName>_ROOT environment variable, where <PackageName> is the package to be found. The package root variables are maintained as a stack so if called from within a find module, root paths from the parent's find module will also be searched after paths for the current package. This can be skipped if NO_PACKAGE_ROOT_PATH is passed or by setting the CMAKE_FIND_USE_PACKAGE_ROOT_PATH to FALSE. See policy CMP0074.

**Comment by byrnHDF** [comment]
Using the HDF5Examples project (included with hdf5 binaries) I was able to test this with a system CMake > 3.12.
Using cmake minimum of 3.12 in the CMakeLists.txt file and defining HDF5_ROOT, the package was found for 1.10.7 or 1.10.8.
Using cmake minimum less 3.12 in the CMakeLists.txt file and enabling policy CMP0074 and defining HDF5_ROOT,  the package was found for 1.10.7 or 1.10.8.

**Comment by jbigot** [comment]
Indeed setting `HDF5_ROOT` is an option. However, HDF5 will not be "automatically detected" as per usual with cmake because the `hdf5*` part must be in the path leading to the file in addition to the file name itself. Also, cmake 3.12 becomes a dependency not only to compile HDF5, but also to consume it.

moving everything under a `hdf5` sub-directory would fulfill the users and cmake's expectations, however I will not enter a fight for it :) If you insist on keeping the current behavior, I guess you can close the issue. It would maybe be nice however to document the fact that `HDF5_ROOT` *must* be specified in order to find HDF5.

**Comment by byrnHDF** [comment]
Yes, we plan to add a FAQ paragraph with the info above and add that info to the USING_HDF5_CMake.txt and USING_CMake_Examples.txt files as well.

**Comment by byrnHDF** [comment]
Welcome to comment on the PR#1414.

**Comment by jbigot** [comment]
Did you make your test under Unix?

From what I understand from cmake doc:
* `HDF5_ROOT` adds an element to the "set of installation prefixes" (later `<prefix>`)
* on windows, the following path are searched for:
  1. `<prefix>/`
  2. `<prefix>/(cmake|CMake)/`
  3. `<prefix>/<name>*/`
  4. `<prefix>/<name>*/(cmake|CMake)/`
  5. `<prefix>/<name>*/(lib/<arch>|lib*|share)/cmake/<name>*/`
  6. `<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/`
  7. `<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/`
* on unix, the following path are searched for:
  1. `<prefix>/(lib/<arch>|lib*|share)/cmake/<name>*/`
  2. `<prefix>/(lib/<arch>|lib*|share)/<name>*/`
  3. `<prefix>/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/`
  4. `<prefix>/<name>*/(lib/<arch>|lib*|share)/cmake/<name>*/`
  5. `<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/`
  6. `<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/`

So, `${HDF5_ROOT}/share/cmake/hdf5-config.cmake` should be found on windows by setting `HDF5_ROOT` to `${HDF5_INSTALL_ROOT}/share/cmake` (case i.) or `${HDF5_INSTALL_ROOT}/share` (case ii.). This is not the usual behavior (cmake intends `HDF5_ROOT` to be the root of the install) but it should work.

On unix however, I don't see how one could make any of the 6 options match `${HDF5_ROOT}/share/cmake/`, option ii. for example would match `${HDF5_ROOT}/share/cmake/hdf5-whatever/hdf5-config.cmake`, but not `${HDF5_ROOT}/share/cmake/hdf5-config.cmake`.

**Comment by byrnHDF** [comment]
Yes linux, you set HDF5_ROOT to the install dir; as used in the previous reply that would be `<prefix>` .
Also, I used HDF5_DIR set to `<prefix>/share/cmake` but that only looks in that one folder.

I still think you are misinterpreting the `<name>*` as a folder, it is the files.
In addition, the docs state that all the default paths are searched; U and W on all platforms.

**Comment by jbigot** [comment]
I do indeed interpret this list as a list of directories. The list is introduced in the doc by:
>  The tables below show the directories searched.

Later:
> This set of directories is intended to work in cooperation with projects that provide configuration files in their installation trees

And each entry is terminated by a slash.



In addition, before the documentation states:
> Package maintainers providing CMake package configuration files are encouraged to name and install them such that the Config Mode Search Procedure outlined below will find them without requiring use of additional options.

**Comment by byrnHDF** [comment]
Looking at the CMake source and with my limited understanding, one of the searches in the code seems to be searching the cmake directory for files matching the project name. But I would prefer someone more experienced look at it.

In reference to the last statement about "without requiring use of additional options", that is what was failing for community members until we removed the "hdf5" directory.

**Comment by jbigot** [comment]
This is strange according to the doc for unix and according to my experience. Isn't this an issue of system difference where the file should be installed to a different path depending on the system family?

**Comment by ax3l** [comment]
Hi y'all,

Let me clarify what that doc page means.
The only hint to find an installed package is the path to its **install prefix**. That's why that doc page shows a series of acceptable **internal** structures _inside_ that prefix, so that you can put your configs somewhere "deeper" in the tree, according to your needs.

The environment hints in modern CMake are either `<CaseSensitivePackageName>_ROOT` or `CMAKE_PREFIX_PATH` and should work when one gives them the _exact same path_ as you used during install for [-DCMAKE_INSTALL_PREFIX=...](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#cmake-install-prefix).

### Example (Unix)

```bash
cmake -S . -B build -DCMAKE_INSTALL_PREFIX=/my/stuff/hdf5-1.13.0
cmake --build build --target install
```

usage:
```bash
export HDF5_ROOT=/my/stuff/hdf5-1.13.0
# or
export CMAKE_PREFIX_PATH=/my/stuff/hdf5-1.13.0;/some/other/stuff/;/more/software

# for a project that searches for HDF5 via `find_package`:
cmake -S . -B build
```

### Example (Windows)

```bat
cmake -S . -B build -G "Ninja" -DCMAKE_INSTALL_PREFIX="C:\\Program Files\HDF5-1.13.0"
cmake --build build --target install
```

usage:
```bat
set "HDF5_ROOT=C:\\Program Files\HDF5-1.13.0"
# or
set "CMAKE_PREFIX_PATH=C:\\Program Files\HDF5-1.13.0;C:\\Program Files\stuff-0.12;C:\\Program Files\foobar-1.2.3"

# for a project that searches for HDF5 via `find_package`:
cmake -S . -B build -G "Ninja"
```

(I picked an arbitrary generator `-G` for Windows for the example, works the same with the Visual Studio one.)

**Comment by jbigot** [comment]
Thank you for the clarification @ax3l . However in this case, the question was more regarding the second part of the path. Where should HDF5 install its files so that they can be found by setting the right prefix?

My understanding is that by installing to:
`${CMAKE_INSTALL_PREFIX}/share/cmake/hdf5-config.cmake`
no single prefix can be set to find this file under unix because the second part of the path matches none of the possible ones.

**Comment by byrnHDF** [comment]
> Hi y'all,
> 
> Let me clarify what that doc page means. The only hint to find an installed package is the path to its **install prefix**. That's why that doc page shows a series of acceptable **internal** structures _inside_ that prefix, so that you can put your configs somewhere "deeper" in the tree, according to your needs.
> 
> The environment hints in modern CMake are either `<CaseSensitivePackageName>_ROOT` or `CMAKE_PREFIX_PATH` and should work when one gives them the _exact same path_ as you used during install for [-DCMAKE_INSTALL_PREFIX=...](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#cmake-install-prefix).
> 
> ### Example (Unix)
> ```shell
> cmake -S . -B build -DCMAKE_INSTALL_PREFIX=/my/stuff/hdf5-1.13.0
> cmake --build build --target install
> ```
> 
> usage:
> 
> ```shell
> export HDF5_ROOT=/my/stuff/hdf5-1.13.0
> # or
> export CMAKE_PREFIX_PATH=/my/stuff/hdf5-1.13.0;/some/other/stuff/;/more/software
> 
> # for a project that searches for HDF5 via `find_package`:
> cmake -S . -B build
> ```
> 
> ### Example (Windows)
> ```batchfile
> cmake -S . -B build -G "Ninja" -DCMAKE_INSTALL_PREFIX="C:\\Program Files\HDF5-1.13.0"
> cmake --build build --target install
> ```
> 
> usage:
> 
> ```batchfile
> set "HDF5_ROOT=C:\\Program Files\HDF5-1.13.0"
> # or
> set "CMAKE_PREFIX_PATH=C:\\Program Files\HDF5-1.13.0;C:\\Program Files\stuff-0.12;C:\\Program Files\foobar-1.2.3"
> 
> # for a project that searches for HDF5 via `find_package`:
> cmake -S . -B build -G "Ninja"
> ```
> 
> (I picked an arbitrary generator `-G` for Windows for the example, works the same with the Visual Studio one.)

I think that explanation would make a great FAQ on our site!

**Comment by byrnHDF** [comment]
I was wrong and the config files will now be in the cmake sub-folder - The previous HDF5_ROOT and CMAKE_PREFIX_PATH are still valid and has been tested to work.

**Comment by ax3l** [comment]
Hi @jbigot, thanks for the kind words:

> Where should HDF5 install its files so that they can be found by setting the right prefix?

The easiest place are the ones [you listed above](https://github.com/HDFGroup/hdf5/issues/1320#issuecomment-1026056375) (and [here](https://github.com/HDFGroup/hdf5/issues/1320#issuecomment-1007355837)).

> My understanding is that by installing to:
> ${CMAKE_INSTALL_PREFIX}/share/cmake/hdf5-config.cmake
> no single prefix can be set to find this file under unix because the second part of the path matches none of the possible ones.

Yep, that one is only for unix: `<prefix>/share/cmake/`

Instead use one of the ones [you listed above](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure):
```
<prefix>/<name>*/(lib/<arch>|lib*|share)/cmake/<name>*/         (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/               (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/ (W/U)
```

Of code an `if` in the CMake install logic to use different ones for Unix and Windows when copying the files into the install tree.
I would go simple and use one of the the three schemes quoted here, so you don't need to branch for the two operating systems. The hints `HDF5_ROOT` (env), `HDF5_DIR` (var), and `CMAKE_PREFIX_PATH` (env) then just point to `<prefix>` and you are done. If `<prefix>` is a system path, e.g., `/usr` then no hints are needed.

**Comment by ax3l** [comment]
@byrnHDF thanks for the feedback! Is there a new related PR that fixes this in `develop`? (Releases 1.10.8/1.13.0 were broken.)

> I was wrong and the config files will now be in the cmake sub-folder - The previous HDF5_ROOT and CMAKE_PREFIX_PATH are still valid and has been tested to work.

The previous ones should work for Unix but not yet for Windows. As in the previous message: let's use an `(W/U)` path in the install tree we create?

Along these lines: `HDF5_INSTALL_CMAKE_DIR=hdf5/share/cmake/hdf5` instead of the Unix-only `share/cmake/hdf5`

**Comment by byrnHDF** [comment]
I have merged and tested (multiple ways) the current code in the development branches.
The W/U paths all will be searched, just that they are all dirs (the * meaning is on the dir and not files as I thought)
the simple path <prefix>/cmake will work.
